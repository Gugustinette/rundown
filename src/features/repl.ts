/**
 * Mostly copied from https://github.com/privatenumber/tsx/blob/master/src/patch-repl.ts
 */
import repl, { type REPLServer, type REPLEval } from "node:repl";
import oxc from "oxc-transform";

const patchEval = (nodeRepl: REPLServer) => {
	const { eval: defaultEval } = nodeRepl;
	const preEval: REPLEval = async function (code, context, filename, callback) {
		let codeToRun = code;
		try {
			const { code: transformedCode } = oxc.transform("index.ts", code, {
				target: "esnext",
				define: {
					require: "global.require",
				},
			});
			codeToRun = transformedCode;
		} catch {
			console.error(
				"Failed to transform code with oxc. Consider opening an issue at https://github.com/Gugustinette/rundown/issues with the code you tried to run.",
			);
		}

		return defaultEval.call(this, codeToRun, context, filename, callback);
	};

	// @ts-expect-error overwriting read-only property
	nodeRepl.eval = preEval;
};

const { start } = repl;
repl.start = function (...args) {
	const nodeRepl = Reflect.apply(start, this, args);
	patchEval(nodeRepl);
	return nodeRepl;
};
