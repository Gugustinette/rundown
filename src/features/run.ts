import { rolldown } from "rolldown";
import { execute } from "../utils/execute";

/**
 * Run a given TypeScript file using Rolldown.
 * @param {string} filePath - The path to the TypeScript file to run.
 * @param {...any} args - Additional arguments to pass to the Node process.
 * @returns {Promise<string>} - A promise that resolves with the code's output when the file has finished running.
 */
// biome-ignore lint/suspicious/noExplicitAny: Any is used here to allow flexibility in the arguments passed to the Node process
export const run = async (filePath: string, ...args: any): Promise<string> => {
	// Setup bundle
	const bundle = await rolldown({
		// Input options (https://rolldown.rs/reference/config-options#inputoptions)
		input: filePath,
	});

	// Generate bundle in memory
	const rolldownOutput = await bundle.generate({
		// Output options (https://rolldown.rs/reference/config-options#outputoptions)
		format: "esm",
	});

	// Verify that the output is not empty
	if (!rolldownOutput.output[0]) {
		throw new Error("No output chunk found");
	}
	// Get the output chunk
	const outputChunk = rolldownOutput.output[0];
	const code = outputChunk.code;

	// Run the code
	const result = await execute(code, ...args);
	// Return the output
	return result.output;
};
