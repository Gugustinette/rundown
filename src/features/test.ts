import { tap } from "node:test/reporters";
import { run } from "node:test";
import process from "node:process";
import path from "node:path";

/**
 * Extend the Node.js test runner to support TypeScript files.
 */
export const test = async (): Promise<void> => {
	return new Promise<void>((resolve, reject) => {
		run({ files: [path.resolve(process.cwd(), "playground/test.test.mjs")] })
			.compose(tap)
			.pipe(process.stdout)
			.on("close", (exitCode) => {
				if (exitCode === 0) {
					resolve();
				} else {
					reject(new Error(`Test suite failed with exit code ${exitCode}`));
				}
			})
			.on("error", (error) => {
				reject(new Error(`Failed to run test suite: ${error.message}`));
			});
	});
};
