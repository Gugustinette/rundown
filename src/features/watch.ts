import { watch as rolldownWatch } from "rolldown";
import { run } from "./run";

/**
 * Run a given TypeScript file using Rolldown.
 * @param {string} filePath - The path to the TypeScript file to run.
 * @returns {Promise<void>} - A promise that resolves when the watcher has finished running.
 */
export const watch = async (filePath: string): Promise<void> => {
	const watcher = rolldownWatch({
		input: filePath,
		output: {
			format: "esm",
		},
		watch: {
			// Prevent the watcher from writing the output to disk
			skipWrite: true,
		},
	});

	watcher.on("event", async (event) => {
		if (event.code === "START") {
			// Clear the console
			console.clear();
			// Run the file using Rolldown
			await run(filePath);
		}
	});

	// Wait for the user to press Ctrl+C (uses await)
	// This is a workaround to keep the process running
	// while the watcher is active
	await new Promise<void>((resolve) => {
		process.on("SIGINT", () => {
			watcher.close();
			resolve();
		});
	});
};
