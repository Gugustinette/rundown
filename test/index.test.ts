import { expect, test, describe } from "vitest";
import { run } from "../src/features/run";

describe("rundown", () => {
	test("should run hello world", async () => {
		const result = await run("./test/fixtures/hello-world.data.ts");
		expect(result).toMatchSnapshot();
	});

	test("should run export function hello world", async () => {
		const result = await run("./test/fixtures/export-function.data.ts");
		expect(result).toMatchSnapshot();
	});
});
