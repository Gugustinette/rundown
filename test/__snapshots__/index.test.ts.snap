// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rundown > should run export function hello world 1`] = `
"Hello, world!
"
`;

exports[`rundown > should run hello world 1`] = `
"Hello, world!
"
`;
